atd-etl/app/process/helpers_request_download.py

```diff
@@ -0,0 +1,198 @@
+"""
+CRIS - Request Download Helper
+Author: Austin Transportation Department, Data and Technology Office
+
+Description: The purpose of this script is to provide helper functions
+that can list and download emails from S3, moving files within the S3
+to other folders, parsing emails, downloading the files from a link
+within the emails, and extracting the downloaded files.
+
+The application requires:
+    https://pypi.org/project/requests/
+    https://pypi.org/project/boto3/
+    https://pypi.org/project/mail-parser/
+"""
+import subprocess
+import glob
+import ntpath
+import re
+
+#
+# Libraries
+#
+import requests
+import mailparser
+import boto3
+
+from .config import ATD_ETL_CONFIG
+
+#
+# Initialize some configuration variables
+#
+AWS_CRIS_REQUEST_BUCKET_NAME = ATD_ETL_CONFIG["AWS_CRIS_REQUEST_BUCKET_NAME"]
+AWS_CRIS_REQUEST_BUCKET_PATH = ATD_ETL_CONFIG["AWS_CRIS_REQUEST_BUCKET_PATH"]
+AWS_CRIS_REQUEST_DOWNLOAD_LINK_PATTERN = ATD_ETL_CONFIG["ATD_CRIS_REQUEST_DOWNLOAD_LINK_PATTERN"]
+ATD_CRIS_REQUEST_RETRIEVE_URL_ENDPOINT = ATD_ETL_CONFIG["ATD_CRIS_REQUEST_RETRIEVE_URL_ENDPOINT"]
+
+#
+# Let's initialize our client object
+#
+aws_s3_client = boto3.client('s3', aws_access_key_id=ATD_ETL_CONFIG["AWS_ACCESS_KEY_ID"],
+                            aws_secret_access_key=ATD_ETL_CONFIG["AWS_SECRET_ACCESS_KEY"])
+
+
+def aws_list_files():
+    """
+    Returns a list of email files.
+    :return: object
+    """
+    global aws_s3_client
+    response = aws_s3_client.list_objects(Bucket=AWS_CRIS_REQUEST_BUCKET_NAME,
+                                   Prefix=AWS_CRIS_REQUEST_BUCKET_PATH)
+
+    for content in response.get('Contents', []):
+        yield content.get('Key')
+
+
+def get_email_list():
+    """
+    Returns an array of files parsed into an actual array (as opposed to an object)
+    :return: array of strings
+    """
+    email_file_list = []
+    pending_email_list = aws_list_files()
+    for email_file in pending_email_list:
+        email_file_list.append(email_file)
+
+    # Remove the first item, it is not needed
+    # since it is just the name of the folder
+    email_file_list.pop(0)
+
+    # Finally return the final list
+    return email_file_list
+
+
+def get_file_name(file_key):
+    """
+    Returns the name of an email file based on the full s3 file path
+    :param file_key: the file path
+    :return: string
+    """
+    return "email_" + ntpath.basename(file_key)
+
+
+def download_s3_file(file_key):
+    """
+    Downloads an email file from s3
+    :param file_key: the full path to the email file
+    :return:
+    """
+    with open("/app/tmp/%s" % (get_file_name(file_key)), 'wb') as f:
```

Comment:
what does the `wb` mean here?

Reply:
and the syntax of the `%`?

Reply:
wb = write binary, %s = string, % = replace %s with variable